crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,171 @@
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing author into TODO")
+    }
+}
+
+#[violation]
+pub struct MissingLink;
+impl Violation for MissingLink {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("To be implemented")
+    }
+}
+
+#[violation]
+pub struct MissingColonInTodo;
+impl Violation for MissingColonInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing colon in TODO")
+    }
+}
+
+#[violation]
+pub struct MissingTextInTodo;
+impl Violation for MissingTextInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing text in TODO")
+    }
+}
+
+// TODO
+// #[violation]
+// pub struct InvalidTODOCapitalization {
+//     pub tag: String,
+// }
+
+#[violation]
+pub struct MissingSpaceAfterColonInTodo;
+impl Violation for MissingSpaceAfterColonInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing space after colon in TODO")
+    }
+}
+
+// Capture groups correspond to checking:
+//      1. Tag (used to match against `TODO` in capitalization and spelling, e.g. ToDo and FIXME)
+//      2. Author exists (in parentheses) after tag, but before colon
+//      3. Colon exists after author
+//      4. Space exists after colon
+//      5. Text exists after space
+//
+// We can check if any of these exist in one regex. Capture groups that don't pick anything up
+// evaluate to `None` in Rust, so the capture group index will always correspond to its respective rule
+// whether the token has been found or not.
+//
+// Example:
+// ```python
+// # TODO(evanrittenhouse): This is a completely valid TODO
+// ```
+// will yield [Some("TODO") Some("evanrittenhouse"), Some(":"), Some(" "), Some("This is a completely valid TODO")], whereas
+// ```python
+// # ToDo this is completely wrong
+// ```
+// will yield [Some("ToDo"), None, None, Some(" "), Some("this is completely wrong")]. Note the
+// `Nones` for the colon and space checks
+//
+// Note: Tags taken from https://github.com/orsinium-labs/flake8-todos/blob/master/flake8_todos/_rules.py#L12.
+static TODO_REGEX: Lazy<Regex> = Lazy::new(|| {
+    // TODO BEFORE COMMITTING - <space> should be a nested group inside of <colon>
+    Regex::new(r"^#\s*(TODO|BUG|FIXME|XXX)(\(.*\))?(:)?( )?(.+)?$").unwrap()
+});
+static NUM_CAPTURE_GROUPS: usize = 5usize;
+
+pub fn check_rules(tokens: &[LexResult]) -> Vec<Diagnostic> {
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+
+    for (start, token, end) in tokens.iter().flatten() {
+        let Tok::Comment(comment) = token else {
+            continue;
+        };
+
+        if get_captured_matches(comment).peek().is_some() {
+            diagnostics.extend(get_tag_regex_errors(comment, *start, *end));
+        }
+    }
+
+    diagnostics
+}
+
+fn get_captured_matches(text: &str) -> Peekable<CaptureMatches> {
+    TODO_REGEX.captures_iter(text).peekable()
+}
```

Comment:
Using `capture` groups is kind of expensive. Is it necessary to get the captures or is it sufficient to only test if the regex matches?

Reply:
The only two alternatives I could think of were:
1. Write different regexes for each combination of successful/failed rules.
2. Just capture a TODO tag and perform string manipulations on the line (for example, split at a colon and ensure that the next character is a space). This _could _ involve additional regexes, though I'd have to think about it more. 

Neither option felt sufficient to warrant using it over capture groups to me, but happy to change approaches if you'd like. 

Out of curiosity, how do you know that regex captures are expensive?

Reply:
The regex crate documentation has some notes on performance and recommend to only [use what you need](https://github.com/rust-lang/regex/blob/master/PERFORMANCE.md#only-ask-for-what-you-need). That's why I asked if we "need" it, because, if not, it would be an easy performance win. But it's okay to use it if we need it. 

You can see some real-world wins in #3735

Reply:
Oh cool, thanks!

 I think that captures are definitely the easiest solution. Matching is very tough since there could be missing pieces at any point in the comment. I checked the benchmarks above and they seem alright. If it's cool with you, I'm okay to move forward with this approach - if there are any slowdowns in the future, we could revisit this?

Reply:
Sounds good to me!